homeassistant/components/elmo_alarm/__init__.py

```diff
@@ -0,0 +1,239 @@
+    )
+
+    async def update():
+        _LOGGER.debug("Connecting to e-connect to retrieve states")
+        await client.update()
+        async_dispatcher_send(hass, SIGNAL_ARMING_STATE_CHANGED, client.state)
+
+        for zone in client.zones:
+            async_dispatcher_send(hass, SIGNAL_ZONE_CHANGED, zone)
+
+        for inp in client.inputs:
+            async_dispatcher_send(hass, SIGNAL_INPUT_CHANGED, inp)
+
+        async_call_later(
+            hass,
+            scan_interval.total_seconds(),
+            lambda _: hass.async_create_task(update()),
+        )
+
+    await update()
+
+    return True
+
+
+class ElmoClientWrapper(ElmoClient):
+    """Wrapping the Elmo client class to adapt it to home assistant."""
+
+    def __init__(self, host, vendor, username, password, states_config):
+        """Initialize the elmo client wrapper."""
+
+        self._username = username
+        self._password = password
+        self._states_config = states_config
+        self._data = None
+        self.state = None
+        self.states = None
+        self.zones = None
+        self.inputs = None
+
+        ElmoClient.__init__(self, host, vendor)
+
+    async def update(self):
+        """Get updates and refresh internal states."""
+        try:
+            data = self.check()
+        except PermissionDenied as exception:
+            _LOGGER.warning("Invalid session, trying to authenticate %s", exception)
+            try:
+                self.auth(self._username, self._password)
+                data = self.check()
+            except PermissionDenied as exception:
+                _LOGGER.warning("Invalid credentials: %s", exception)
+            except HTTPError as exception:
+                _LOGGER.warning(
+                    "Got HTTP error when authenticating. Check credentials. Code: %s",
+                    exception,
+                )
+
+        if self._data is None:
+            await self._configure_states()
+        self._data = data
+        await self._update_arm_state()
+        await self._update_zone_state()
+        await self._update_input_state()
+
+    async def _configure_states(self):
+        """Initialize the elmo alarm states."""
+        state_config = {}
+        for state in self._states_config:
+            state_config[state[CONF_NAME]] = state[CONF_ZONES]
+        self.states = state_config
+
+    async def _update_arm_state(self):
+        """Update the elmo alarm states."""
+        areas_armed = self._data["areas_armed"]
+
+        if not areas_armed:
+            self.state = STATE_ALARM_DISARMED
+        else:
+            armed_indexes = [x["index"] + 1 for x in areas_armed]
+
+            state = [
+                state
+                for state, areas in self.states.items()
+                if set(areas) == set(armed_indexes)
+            ]
+
+            if state:
+                state = state[0]
+                if state == "arm_away":
```

Comment:
Create constant for the Elmo alarm states.

Reply:
Removed custom states reusing states from HA const.